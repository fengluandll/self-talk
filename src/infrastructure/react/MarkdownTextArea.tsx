import React, {
  ChangeEvent,
  ClipboardEvent,
  DragEvent,
  SyntheticEvent
} from "react";
import { InsertImagesFunction } from "./utilities";
import { TextArea } from "./TextArea";

interface IProps {
  className?: string;
  insertImages: InsertImagesFunction;
  onSubmit: () => Promise<void>;
  setText: (text: string) => void;
  text: string;
}

export const MarkdownTextArea = ({
  insertImages,
  onSubmit,
  setText,
  text,
  ...restProps
}: IProps) => {
  const uploadImages = async (
    event: SyntheticEvent<HTMLTextAreaElement>,
    files: Array<File | null>
  ) => {
    const images = files
      .filter((file): file is File => !!file)
      .filter(file => file.type.startsWith("image/"));

    if (images.length === 0) {
      return;
    }

    setText(
      await insertImages(text, event.currentTarget.selectionStart, images)
    );
  };

  return (
    <TextArea
      onSubmit={onSubmit}
      placeholder="Write in Markdown ..."
      onChange={(event: ChangeEvent<HTMLTextAreaElement>) =>
        setText(event.target.value)
      }
      onDrop={(event: DragEvent<HTMLTextAreaElement>): Promise<void> =>
        uploadImages(event, Array.from(event.dataTransfer.files))
      }
      onPaste={async (
        event: ClipboardEvent<HTMLTextAreaElement>
      ): Promise<void> => {
        if (!event.clipboardData || !event.clipboardData.items) {
          return;
        }

        await uploadImages(
          event,
          Array.from(event.clipboardData.items).map(item => item.getAsFile())
        );
      }}
      value={text}
      {...restProps}
    />
  );
};
