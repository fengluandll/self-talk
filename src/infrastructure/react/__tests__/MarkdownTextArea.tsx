import React from "react";
import { create } from "react-test-renderer";
import { MarkdownTextArea } from "../MarkdownTextArea";

it("renders", () => {
  expect(
    create(
      <MarkdownTextArea
        insertImage={async () => "url"}
        onSubmit={async () => undefined}
        setText={() => undefined}
        text="foo"
      />
    ).toJSON()
  ).toMatchSnapshot();
});
