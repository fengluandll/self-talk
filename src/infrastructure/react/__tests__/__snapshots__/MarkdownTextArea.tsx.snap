// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<textarea
  async={true}
  className="sc-bdVaJa cPXXOd"
  onChange={[Function]}
  onKeyDown={[Function]}
  onPaste={[Function]}
  placeholder="Write in Markdown ..."
  rows={1}
  value="foo"
/>
`;
